[FLINK-2097] Implement job session management

mxm opened the pull request:
This is a joint effort by @StephanEwen and me to introduce a session management in Flink. Session are used to keep a copy of the ExecutionGraph in the job manager for the session lifetime. It is important that the ExecutionGraph is not kept around longer because it consumes a lot of memory. Its intermediate results can also be freed. To integrate sessions properly into Flink, some refactoring was necessary. In particular these are:

- JobId is created through the ExecutionEnvironment and passed through
- Sessions can be termined by the ExecutionEnvironment or directly through the executor
- Session are cancelled implicitly through "reapers" or shutdown hooks in the ExecutionEnvironment, otherwise they time out
- LocalExecutor and RemoteExecutor manage sessions
- The Client only deals with the communication with the job manager and is agnostic of session management

With the session management, we will be able to properly support backtracking of produced intermediate results. This makes calls to count()/collect()/print() efficient and enables to write incremental/interactive jobs.

uce commented:
I just had a look at the JobManager in a different context and thought about the following, which might be relevant here: when submitting a new JobGraph, which is attached to an existing ExecutionGraph, some ExecutionGraph state is overwritten by the new JobGraph. With some you might run into (maybe) unexpected behaviour like resetting number of left execution retries or creating a new CheckpointCoordinator for the ExecutionGraph.

What's the intended behaviour of attaching to an existing ExecutionGraph? Is there an implicit assumption that the existing ExecutionGraph needs to be finished already?

StephanEwen commented:
I think right now, it pretty much behaves as if someone started a new job, with the "grown" execution graph.

mxm commented:
It should just add more nodes to the ExecutionGraph. Existing ones should not be modified. For batch, I think the assumption is that it needs to be finished. For streaming, I could also picture attaching nodes at runtime but this has to be carefully implemented..

mxm commented:
I've ported this pull request to the latest master. It was a lot more work than I anticipated because some classes had diverged significantly and merging them was a bit hard.

Due to some refactoring, the changes have grown quite large again and I know that makes reviewing hard. Despite that, I wouldn't delay merging this pull request much further. We can disable the session management until it is integrated with the rest of the system (intermediate results) by throwing an exception on the interface methods. If we decide later, that we want to delay this feature, we could also remove the session code. In that case, it would still make sense to merge this pull request because it contains a lot of nice refactoring.

With the session management in place, we can reuse already computed intermediate results with not too much effort. Actually, only some API changes are remaining to expose the session management to the user in production.

rmetzger commented on flink-tests/src/test/scala/org/apache/flink/api/scala/runtime/jobmanager/JobManagerFailsITCase.scala:
```diff
@@ -53,41 +54,41 @@ class JobManagerFailsITCase(_system: ActorSystem)
-        cluster.stop()
-      }
-    }
+//    "detect a lost connection to the JobManager and try to reconnect to it" in {
```
Re-enable the test?

mxm replied:
Thanks.

tillrohrmann commented:
Could you elaborate a little bit on what you refactored and which components would be important to review?

mxm commented:
Of course! The following classes have been refactored in the course of integrating them with the session management:

**Client**
- Establish connection to JobManager on creation
- Refactor run method into `runBlocking` and `runDetached`
- Extract helper classes to generate the Plan
- Make Optimizer and JobGraph generation methods `static`
- Pass `ClassLoader` correctly (do not keep one per Client but rather let it be passed before submission)

**CliFrontend**
- `runBlocking` and `runDetached` methods by analogy with the Client class

**ExecutionEnvironment**, **LocalEnvironment**, **RemoteEnvironment**
- modified abstract class to support sessions (timeout and jobID generation)
- handle session management via Reapers and ShutdownHooks

**PlanExecutor**, **LocalExecutor**, **RemoteExecutor**
- modified interface
- support session termination
- set JobID on Plan

**JobManager**
- keep ExecutionGraph as long as session has not expired

Future issues:
- Support for sessions in streaming. Currently streaming jobs are agnostic of sessions.
- Representation of sessions in the JobManager web frontend. How do we represent updates to the ExecutionGraph in sessions?
- Build features on top of session management (e.g. intermediate results)

tillrohrmann commented:
Thanks Max for the detailed description.

On Wed, Sep 9, 2015 at 11:12 AM, Max <notifications@github.com> wrote:

> Of course! The following classes have been refactored in the course of
> integrating them with the session management:
>
> *Client*
>
>    - Establish connection to JobManager on creation
>    - Refactor run method into runBlocking and runDetached
>    - Extract helper classes to generate the Plan
>    - Make Optimizer and JobGraph generation methods static
>    - Pass ClassLoader correctly (do not keep one per Client but rather
>    let it be passed before submission)
>
> *CliFrontend*
>
>    - runBlocking and runDetached methods by analogy with the Client class
>
> *ExecutionEnvironment*, *LocalEnvironment*, *RemoteEnvironment*
>
>    - modified abstract class to support sessions (timeout and jobID
>    generation)
>    - handle session management via Reapers and ShutdownHooks
>
> *PlanExecutor*, *LocalExecutor*, *RemoteExecutor*
>
>    - modified interface
>    - support session termination
>    - set JobID on Plan
>
> *JobManager*
>
>    - keep ExecutionGraph as long as session has not expired
>
> Future issues:
>
>    - Support for sessions in streaming. Currently streaming jobs are
>    agnostic of sessions.
>    - Representation of sessions in the JobManager web frontend. How do we
>    represent updates to the ExecutionGraph in sessions?
>    - Build features on top of session management (e.g. intermediate
>    results)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/flink/pull/858#issuecomment-138848287>.
>

mxm commented:
I've rebased again...If nobody objects, I will merge this soon. The new API-facing methods on `ExecutionEnvironment` will be disabled until we implement first applications of session management. I've added a separate commit that does that.
